### ISSUE TITLE
Problems using fairseq-generate in fully sharded model (FSDP)

### ISSUE THREAD

**Issue by Zichen-Wu** [facebookresearch/fairseq#4275]
#### Code
In training stage
`
fairseq-train $DATA_DIR \
--ddp-backend fully_sharded --distributed-world-size 2 \
--restore-file $PRETRAINED_MODEL \
--task translation_mbart --user-dir $USER_DIR --arch $ARCH \
--max-tokens 10240 \
--encoder-normalize-before --decoder-normalize-before \
--source-lang src --target-lang tgt \
--langs $langs --task-lang $task_lang \
--use-sharded-state \
--checkpoint-activations \
--layernorm-embedding \
--share-all-embeddings \
--share-decoder-input-output-embed \
--reset-optimizer --reset-dataloader --reset-meters \
--criterion label_smoothed_cross_entropy \
--label-smoothing 0.1 \
--dropout 0.1 --attention-dropout 0.1 \
--weight-decay 0.01 --optimizer adam --adam-betas "(0.9, 0.998)" --adam-eps 1e-08 \
--max-source-positions 1024 --max-target-positions 1024 \
--clip-norm 0.1
--lr-scheduler polynomial_decay --lr 3e-05 --total-num-update 8000 --warmup-updates 500
--fp16 --batch-size 24
--patience 2
--keep-last-epochs 1
--save-dir $SAVE_DIR
--find-unused-parameters
--log-file $SAVE_DIR/train.log
`

Through training, I get finetuned sharded models 'checkpoint_best-shard0.pt' and 'checkpoint_best-shard1.pt'

Then I want to generate using the two models.

`fairseq-generate $DataDir \
--path $MODEL_DIR/checkpoint_best.pt \
--user-dir $USER_DIR --task translation_mbart \
--gen-subset test \
--prefix-size 2 \
--bpe 'sentencepiece' --sentencepiece-model $Model_Dir/sentence.bpe.model \
--sacrebleu --remove-bpe 'sentencepiece' \
--langs $langs --task-lang $task_lang \
--beam 5 --num-workers 4 --no-repeat-ngram-size 3 --lenpen 1.5 2>&1 > $TempFile`

And it failed and have the problems reported:
` Unexpected key(s) in state_dict: "flat_param_0", "encoder.layers.0.flat_param_0", "encoder.layers.1.flat_param_0", "encoder.layers.1.flat_param_0", "encoder.layers.2.flat_param_0 ....."`

#### What have you tried?
I have tried to replace the path arguments 'checkpoint_best.pt' with  'checkpoint_best-shard0.pt:checkpoint_best-shard1.pt', but it did not work. I also want to find some generate scripts using sharded model but unfortunately I didn't find any.

#### What's your environment?

 - fairseq Version (e.g., 1.0 or main): main
 - PyTorch Version (e.g., 1.0) 1.11.0
 - OS (e.g., Linux): Linux
 - Python version: 3.8
 - CUDA/cuDNN version: 11.6
 - GPU models and configuration: NVIDIA GeForce RTX 2080 Ti
 - Any other relevant information: fairscale also main version

**Comment by gmryu** [comment]
I haven't used sharded model, so this might be no helps.
However, I believe you should add more arguments to your generate-command.
Like `--ddp-backend fully_sharded` to make it support flat_param class,
`--checkpoint-shard-count 2` to make fairseq notice there are 2 shards of the same checkpoint.

**Comment by Xingxl2studious** [comment]
i have the same question
